src/widget.ts

```diff
@@ -60,26 +122,44 @@ class CodeMirrorWidget extends Widget {
   /**
    * Construct a CodeMirror widget.
    */
-  constructor(model: IEditorViewModel) {
+  constructor(model: IEditorModel) {
     super();
     this.addClass(FILE_BROWSER_CLASS);
-    this._model = model;
     this._editor = CodeMirror(this.node);
-    this.updateMimetype(model.mimetype);
-    this.updateFilename(model.filename);
-    this.updateReadOnly(model.readOnly);
-    this.updateTabSize(model.tabSize);
-    this.updateLineNumbers(model.lineNumbers);
-    this.updateFixedHeight(model.fixedHeight);
-    this.updateText(model.text);
+    this.model = model;
+  }
+
+  /**
+   * Get the editor model.
+   */
+  get model(): IEditorModel {
+    return this._model;
+  }
 
+  /**
+   * Set the editor model, can not be `null`.
+   */
+  set model(value: IEditorModel) {
+    if (value === null) {
+      return;
+    }
+    if (this._model !== null) {
+      this._model.stateChanged.disconnect(this.onModelStateChanged, this);
+    }
+    this._model = value;
+    this.updateMimetype(value.mimetype);
+    this.updateFilename(value.filename);
+    this.updateReadOnly(value.readOnly);
+    this.updateTabSize(value.tabSize);
+    this.updateLineNumbers(value.lineNumbers);
+    this.updateFixedHeight(value.fixedHeight);
+    this.updateText(value.text);
     this._editor.on('change', (instance, change) => {
```

Comment:
this `on('change'` listener should be installed on the constructor and call into a private method.